[FEATURE] Added iterations to most parts of the benchmark + Memory usage

jules552 opened the pull request:
So here the implementation for adding `--iter NB_OF_ITERATIONS` as an optional argument to the benchmark.

For the moment I've set the default to 1 iteration as asked but I'm not sure that ideal, maybe 5 or 10 would be more representative of an average result.
I haven't added the iteration of the init part for two reason:

1. I don't really know how to scramble/flush all the caches involved such that the time is representative of a fresh initialization. That's one reason why the benchmark of my processor has an init time so low, it seems that most of the model was kept in memory instead of reloaded from the main storage (and even in this case, there is also the DRAM cache of the SSD that would be involved), so it is complicated to fix all of that.

2. It would spam the terminal to do so with the messages generated by llama.cpp at initialization, and I have no idea how to prevent that without modifying the cpp file.

As a bonus, I've added the report of the RAM usage using memory-profiler and added the library as a requirement in the requirement.txt of the project, so now it should be easier for user to instantly see the info they should report. 

There is one cold run before all the iterations such that it fills correctly all the caches such that it would better correspond to the average performance of using this tool. And only the last response of the model is displayed such that it doesn't really spam the terminal output.

If you have anything you want me to change, just ask and will do my best to accommodate such that it can be merged !

liltom-eth commented on benchmark.py:
```diff
@@ -58,36 +67,43 @@ def main():
+    run_iteration(llama2_wrapper, prompt_example, DEFAULT_SYSTEM_PROMPT, DEFAULT_MAX_NEW_TOKENS)
+
+    # Timed runs
+    print("Performing timed runs...")
```
Consider adding  `iter` number in the printout like ` print(f"Performing {args.iter} timed runs...")` .

liltom-eth commented on benchmark.py:
```diff
@@ -58,36 +67,43 @@ def main():
+
+    # Timed runs
+    print("Performing timed runs...")
+    for _ in range(args.iter):
```
Consider add `tqdm` like `for _ in tqdm(range(args.iter)):`
Also `from tqdm import tqdm` as well.

jules552 replied:
Really great library to convey progression to the user, good to know !

liltom-eth reviewed (commented):
Nice work!
Thanks for your contribution.
1.  RAM usage is perfect.
2.  leaving `init` alone out of iteration should be great. Most users only care about inference speed (tokens/sec) on their devices.
3. Cold run is nice and necessary for the benchmark.
4. I prefer to set the default `iter` as 1 since it is faster and can be used as a fast method to check the env installation.

Add some minor suggestions, and PR should be ready for merge after your updates.

liltom-eth commented on benchmark.py:
```diff
@@ -1,43 +1,52 @@
 
 def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--iter", type=int, default=1, help="Number of iterations")
```
yeah, `default=5` should be better.

jules552 commented:
> Nice work! Thanks for your contribution. Tested on my devices and looks great.
> 
>     1. RAM usage is perfect.
> 
>     2. leaving `init` alone out of iteration should be great. Most users only care about inference speed (tokens/sec) on their devices.
> 
>     3. Cold run is nice and necessary for the benchmark.
> 
>     4. Setting the default `iter` as 5 is a good idea. And it can be used as a fast method to check the env installation.
> 
>     5. llama.cpp spamming the terminal, this issue maybe can be solved by setting some arguments for llama.cpp to make it silent in the following PRs.
> 
> 
> Add some minor suggestions, and PR should be ready for merge after your updates. It should be great also add the usage of your work in [readme](https://github.com/liltom-eth/llama2-webui#benchmark).

I've adjusted everything and updated the README accordingly

liltom-eth reviewed (approved):
LGTM.
